src/snipster/repo.py

```diff
@@ -1,99 +1,210 @@
+        self.session.add(stored_snippet)
         self.session.commit()
-        self.session.refresh(snippet_orm)
-        return to_public(snippet_orm)
+        self.session.refresh(stored_snippet)
+        return stored_snippet
 
-    def get(self, snippet_id) -> SnippetPublic | None:
-        snippet_orm = self.session.get(SnippetORM, snippet_id)
-        if snippet_orm:
-            return to_public(snippet_orm)
+    def get(self, snippet_id) -> Snippet:
+        snippet = self.session.get(Snippet, snippet_id)
+        if snippet:
+            return snippet
         raise SnippetNotFoundError
 
-    def list(self) -> Sequence[SnippetPublic]:
-        snippet_orms = self.session.exec(select(SnippetORM)).all()
-        return [to_public(orm) for orm in snippet_orms]
+    def list(self) -> Sequence[Snippet]:
+        snippets = self.session.exec(select(Snippet)).all()
+        return [snippet for snippet in snippets]
 
     def delete(self, snippet_id: int):
-        snippet = self.session.get(SnippetORM, snippet_id)
+        snippet = self.session.get(Snippet, snippet_id)
         if snippet:
             self.session.delete(snippet)
             self.session.commit()
 
+    def toggle_favorite(self, snippet_id: int) -> None:
+        snippet = self.session.get(Snippet, snippet_id)
+        if not snippet:
+            raise SnippetNotFoundError(f"Snippet with id {snippet_id} not found.")
+        snippet.favorite = not snippet.favorite
+        self.session.commit()
+        self.session.refresh(snippet)
+
+    def add_tag(self, snippet_id: int, tag: str) -> None:
+        snippet = self.session.get(Snippet, snippet_id)
+        if not snippet:
+            raise SnippetNotFoundError(f"Snippet with id {snippet_id} not found.")
+        if tag not in snippet.tags:
+            snippet.tags.append(tag)
+            self.session.commit()
+            self.session.refresh(snippet)
+
+    def remove_tag(self, snippet_id: int, tag: str) -> None:
+        snippet = self.session.get(Snippet, snippet_id)
+        if not snippet:
+            raise SnippetNotFoundError(f"Snippet with id {snippet_id} not found.")
+        if tag not in snippet.tags:
+            raise ValueError(f"Tag {tag} not found on snippet with id {snippet_id}.")
+        snippet.tags.remove(tag)
+        self.session.commit()
+        self.session.refresh(snippet)
+
+    def search(self, query: str) -> Sequence[Snippet]:
+        stmt = select(Snippet).where(
+            or_(
+                Snippet.title.ilike(f"%{query}%"),  # type: ignore
+                Snippet.code.ilike(f"%{query}%"),  # type: ignore
+                Snippet.description.ilike(f"%{query}%"),  # type: ignore
+            )
+        )
+        string_search_results = self.session.exec(stmt).all()
+
+        all_snippets = self.session.exec(select(Snippet)).all()
+        tag_search_results = [
+            snippet for snippet in all_snippets if query in snippet.tags
```

Comment:
+ we do Python filtering, let's see if we can refactor this to use SQL / the ORM, e.g. `.where(Snippet.tags.contains(...))`